apps/mark-scan/accessible-controller/src/controllerd.rs

```diff
@@ -0,0 +1,408 @@
+        Ok(_) => println!("Echo command sent"),
+        Err(error) => eprintln!("{error:?}"),
+    }
+
+    let mut serial_buf: Vec<u8> = vec![0; 20];
+
+    let start_time = Instant::now();
+    loop {
+        match port.read(serial_buf.as_mut_slice()) {
+            Ok(size) => {
+                let echo_response = &serial_buf[..size];
+                assert_eq!(
+                    echo_command, echo_response,
+                    "Received different response from echo command: {echo_response:x?}"
+                );
+
+                println!("Received valid echo command response");
+                return Ok(());
+            }
+            Err(ref e) if e.kind() == io::ErrorKind::TimedOut => (),
+            Err(e) => eprintln!("Error reading echo response: {e:?}"),
+        }
+
+        if start_time.elapsed() >= MAX_ECHO_RESPONSE_WAIT {
+            break;
+        }
+
+        thread::sleep(POLL_INTERVAL);
+    }
+
+    Err(io::Error::new(
+        io::ErrorKind::TimedOut,
+        "No echo response received",
+    ))
+}
+
+fn create_virtual_device() -> Device {
+    uinput::open(UINPUT_PATH)
+        .unwrap()
+        .name("Accessible Controller Daemon Virtual Device")
+        .unwrap()
+        .event(uinput::event::Keyboard::All)
+        .unwrap()
+        .create()
+        .unwrap()
+}
+
+fn main() {
+    // Open the serial port
+    let port = serialport::new(DEVICE_PATH, DEVICE_BAUD_RATE)
+        .timeout(POLL_INTERVAL)
+        .open();
+
+    println!("Opened controller serial port at {DEVICE_PATH}");
+
+    // Create virtual device for keypress events
+    let mut device = create_virtual_device();
+    println!("Created virtual device");
+
+    // Wait for virtual device to register
+    thread::sleep(Duration::from_secs(1));
+
+    match port {
+        Ok(mut port) => {
+            validate_connection(&mut port).unwrap();
+
+            println!("Receiving data on {DEVICE_PATH} at {DEVICE_BAUD_RATE} baud");
+
+            let mut serial_buf: Vec<u8> = vec![0; 1000];
+            loop {
+                match port.read(serial_buf.as_mut_slice()) {
+                    Ok(size) => match handle_command(&mut device, &serial_buf[..size]) {
+                        Ok(_) => (),
+                        Err(e) => match e {
+                            CommandError::UnsupportedButton(_) => {
+                                println!("[warn] {e}")
+                            }
+                            _ => panic!("Error handling command: {e}"),
+                        },
+                    },
+                    // Timeout error just means no event was sent in the current polling period
+                    Err(ref e) if e.kind() == io::ErrorKind::TimedOut => (),
+                    Err(e) => eprintln!("{e:?}"),
+                }
+            }
+        }
+        Err(e) => {
+            eprintln!(r#"Failed to open "{DEVICE_PATH}". Error: {e}"#);
+            ::std::process::exit(1);
+        }
```

Comment:
It seems like there are a few different ways we handle unknown errors in this series of match blocks:
- panic (L324)
- eprintln and keep looping (L329)
- eprintln and exit 1 (L335)

I just wanted to make sure that was intentional. My instinct would be to crash in all cases unless there's a specific reason to keep going (which maybe could be documented with a comment)

Reply:
I agree with this. Is exiting really the approach you want? I assume we want this daemon to always be running, so it'd be better to log the error and continue unless the error is truly unrecoverable. Have you thought about how this daemon will be run?

Reply:
Yeah agree with this:
1. eprintln and exit should be replaced with panic for unrecoverable errors
2. we should only exit on truly unrecoverable errors

I think in practice this means we only panic when the device can't be opened after `n` retries, which would indicate a hardware failure. Command errors are likely to be recoverable so we should not panic on those.